rs/ic_os/manifest_tool/src/components_parser.rs

```diff
@@ -0,0 +1,127 @@
+use anyhow::{Context, Result};
+use regex::Regex;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+pub const COMPONENTS_PATH: &str = "ic-os/components";
+
+#[derive(Debug)]
+pub struct IcosManifest {
+    pub setupos: Manifest,
+    pub hostos: Manifest,
+    pub guestos: Manifest,
+    pub boundary_guestos: Manifest,
+}
+
+impl IcosManifest {
+    pub fn new(
+        guestos_components: Manifest,
+        hostos_components: Manifest,
+        setupos_components: Manifest,
+        boundary_guestos_components: Manifest,
+    ) -> Self {
+        IcosManifest {
+            guestos: guestos_components,
+            hostos: hostos_components,
+            setupos: setupos_components,
+            boundary_guestos: boundary_guestos_components,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Manifest {
+    pub manifest: Vec<ManifestEntry>,
+}
+
+impl Manifest {
+    pub fn new() -> Self {
+        Manifest {
+            manifest: Vec::new(),
+        }
+    }
+
+    pub fn add_entry(&mut self, component: ManifestEntry) {
+        self.manifest.push(component);
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ManifestEntry {
+    pub source: PathBuf,
+    #[allow(dead_code)]
+    pub destination: PathBuf,
+}
+
+impl ManifestEntry {
+    pub fn new(source: PathBuf, destination: PathBuf) -> Self {
+        ManifestEntry {
+            source,
+            destination,
+        }
+    }
+}
+
+fn get_manifest(manifest_path: &Path, components_path: &Path) -> Result<Manifest> {
+    let manifest_contents = fs::read_to_string(manifest_path)
+        .with_context(|| format!("Failed to read file: {:?}", manifest_path))?;
+
+    let re = Regex::new(r#"Label\("(.+?)"\): "(.+?)""#)?;
+
+    let mut manifest = Manifest::new();
+
+    for cap in re.captures_iter(&manifest_contents) {
+        let source = components_path.join(&cap[1]);
+        let destination = PathBuf::from(&cap[2]);
+        manifest.add_entry(ManifestEntry::new(source, destination));
+    }
+
+    Ok(manifest)
+}
+
+pub fn get_icos_manifest(repo_root: &Path) -> Result<IcosManifest> {
```

Comment:
This could be raised one level up, since it is really just shuffling paths into the `new` call, sort of outside the logic of the parser.

Reply:
Alternatively, I could move this logic into the IcosManifest constructor. WDYT?